File: go/internal/pkgfile/reader.go
package pkgfile

import (
	"encoding/binary"
	"fmt"
	"io"

	"github.com/1m188/repkg-go/internal/binutil"
)

// maxMagicLength PKG magic 最大长度。
const maxMagicLength = 32

// maxFilePathLength 条目路径最大长度。
const maxFilePathLength = 255

// maxEntryCount PKG 条目数量上限（防止恶意文件导致 OOM）。
const maxEntryCount = 100_000

// Reader 用于读取 PKG 文件。
// 对应 C# PackageReader。
type Reader struct {
	// ReadEntryBytes 控制是否读取条目的字节数据。默认 true。
	ReadEntryBytes bool
}

// NewReader 创建默认的 PKG 读取器。
func NewReader() *Reader {
	return &Reader{ReadEntryBytes: true}
}

// ReadPackage 从 reader 中读取并解析 PKG 数据。
//
//nolint:funlen // PKG 头部解析需要多步骤
func (r *Reader) ReadPackage(reader io.Reader) (*Package, error) { //nolint:gocyclo,cyclop // PKG 二进制解析需要多步骤处理
	pkg := &Package{}

	// 读取 magic 字符串
	magic, err := binutil.ReadStringI32Size(reader, maxMagicLength)
	if err != nil {
		return nil, fmt.Errorf("读取 PKG magic 失败: %w", err)
	}
	pkg.Magic = magic

	// 读取条目数量
	var entryCount int32
	err = binary.Read(reader, binary.LittleEndian, &entryCount)
	if err != nil {
		return nil, fmt.Errorf("读取条目数量失败: %w", err)
	}
	if entryCount < 0 {
		return nil, fmt.Errorf("条目数量无效: %d", entryCount)
	}
	if entryCount > maxEntryCount {
		return nil, fmt.Errorf("条目数量超出限制: %d / %d", entryCount, maxEntryCount)
	}

	// 读取条目列表（头部信息）
	entries := make([]*Entry, entryCount)
	for i := range entryCount {
		// 读取完整路径
		fullPath, err := binutil.ReadStringI32Size(reader, maxFilePathLength)
		if err != nil {
			return nil, fmt.Errorf("读取条目路径失败 (索引 %d): %w", i, err)
		}

		// 读取偏移
		var offset int32
		err = binary.Read(reader, binary.LittleEndian, &offset)
		if err != nil {
			return nil, fmt.Errorf("读取条目偏移失败 (索引 %d): %w", i, err)
		}

		// 读取长度
		var length int32
		err = binary.Read(reader, binary.LittleEndian, &length)
		if err != nil {
			return nil, fmt.Errorf("读取条目长度失败 (索引 %d): %w", i, err)
		}

		entry := &Entry{
			FullPath: fullPath,
			Offset:   offset,
			Length:   length,
			Type:     EntryTypeFromFileName(fullPath),
		}
		entries[i] = entry
	}

	pkg.Entries = entries

	// 计算头部大小（magic、entryCount、各条目元数据的总字节数）
	headerSize := int32(4 + len(pkg.Magic) + 4) //nolint:gosec // 头部大小受限于入口数据量
	for _, entry := range entries {
		headerSize += 4 + int32(len(entry.FullPath)) + 4 + 4 //nolint:gosec // 头部大小受限于入口数据量
	}
	pkg.HeaderSize = headerSize

	// 不读取字节数据时直接返回
	if !r.ReadEntryBytes {
		return pkg, nil
	}

	// 读取每个条目的字节数据（通过 Offset 字段定位，而非顺序读取）
	bodyBytes, err := io.ReadAll(reader)
	if err != nil {
		return nil, fmt.Errorf("读取 PKG 数据体失败: %w", err)
	}

	for _, entry := range entries {
		off := entry.Offset
		bodyLen := int64(len(bodyBytes))
		if off < 0 || entry.Length < 0 || int64(off)+int64(entry.Length) > bodyLen {
			return nil, fmt.Errorf("条目数据偏移越界 (%s): offset=%d length=%d bodyLen=%d",
				entry.FullPath, off, entry.Length, len(bodyBytes))
		}
		entry.Bytes = make([]byte, entry.Length)
		copy(entry.Bytes, bodyBytes[off:off+entry.Length])
	}

	return pkg, nil
}

File: rs/src/pkgfile/reader.rs
// PKG 读取器，负责从二进制流中解析 PKG 文件。
// 与 C# 原版 RePKG.Application/Package/PackageReader.cs 对应。

use std::io::{self, Read};

use crate::binutil;

use super::{Entry, EntryType, Package};

/// PKG 文件安全常量。
const MAX_MAGIC_LENGTH: usize = 32;
const MAX_FILE_PATH_LENGTH: usize = 255;
const MAX_ENTRY_COUNT: usize = 100_000;

/// 从 reader 中读取并解析 PKG 文件。
pub fn read_package<R: Read>(reader: &mut R) -> io::Result<Package> {
    let magic = binutil::read_string_i32_size(reader, MAX_MAGIC_LENGTH)?;

    let mut buf = [0u8; 4];
    reader.read_exact(&mut buf)?;
    let entry_count_i32 = i32::from_le_bytes(buf);
    if entry_count_i32 < 0 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("条目数无效（负数）：{}", entry_count_i32),
        ));
    }
    let entry_count = entry_count_i32 as usize;
    if entry_count > MAX_ENTRY_COUNT {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("条目数超限：{}（最大 {}）", entry_count, MAX_ENTRY_COUNT),
        ));
    }

    let mut entries = Vec::with_capacity(entry_count);
    for _ in 0..entry_count {
        let full_path = binutil::read_string_i32_size(reader, MAX_FILE_PATH_LENGTH)?;

        reader.read_exact(&mut buf)?;
        let offset = i32::from_le_bytes(buf);

        reader.read_exact(&mut buf)?;
        let length = i32::from_le_bytes(buf);

        if offset < 0 || length < 0 {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!(
                    "条目 \"{}\" 偏移或长度无效：offset={}, length={}",
                    full_path, offset, length
                ),
            ));
        }

        let entry_type = if full_path.to_lowercase().ends_with(".tex") {
            EntryType::Tex
        } else {
            EntryType::Binary
        };

        entries.push(Entry {
            full_path,
            offset,
            length,
            bytes: Vec::new(),
            entry_type,
        });
    }

    // 读取数据体
    let mut body = Vec::new();
    reader.read_to_end(&mut body)?;

    // 根据 offset 填充每条的 bytes
    for entry in entries.iter_mut() {
        let off = entry.offset as usize;
        let len = entry.length as usize;
        if off + len > body.len() {
            return Err(io::Error::new(
                io::ErrorKind::UnexpectedEof,
                format!(
                    "条目 \"{}\" 数据截断：偏移 {} + 长度 {} > 数据体长度 {}",
                    entry.full_path,
                    off,
                    len,
                    body.len()
                ),
            ));
        }
        entry.bytes = body[off..off + len].to_vec();
    }

    Ok(Package { magic, entries })
}
